proposals/254-padding-negotiation.txt

```diff
@@ -71,331 +77,66 @@ the circuit.
-
-The fixed schedule padding will be encoded in a
-RELAY_COMMAND_PADDING_SCHEDULE cell. It specifies a set of up to 80
-fixed time points in the future to send cells.
-
-XXX: 80 timers is a lot to allow every client to create. We may want to
-have something that checks this structure to ensure it actually
-schedules no more than N in practice, until we figure out how to
-optimize either libevent or timer scheduling/packet delivery. See also
-Section 4.3.
-
-The RELAY_COMMAND_PADDING_SCHEDULE body is specified in Trunnel as
-follows:
-
-    struct relay_padding_schedule {
-       u8 schedule_length IN [1..80];
-
-       /* Number of microseconds before sending cells (cumulative) */
-       u32 when_send[schedule_length];
-
-       /* Number of cells to send at time point sum(when_send[0..i]) */
-       u16 num_cells[schedule_length];
-
-       /* Adaptivity: If 1, and server-originating cells arrive before the
-          next when_send time, then decrement the next non-zero when_send
-          index, so we don't send a padding cell then, too */
-       u8 adaptive IN [0,1];
-    };
+The statistical mechanisms that define padding are known as padding
+machines. Padding machines can be hardcoded in Tor, specified in the
+consensus, and custom research machines can be listed in Torrc.
 
-To allow both high-resolution time values, and the ability to specify
-timeout values far in the future, the time values are cumulative. In
-other words, sending a cell with when_send = [MAX_INT, MAX_INT, MAX_INT,
-0...] and num_cells = [0, 0, 100, 0...] would cause the relay to reply
-with 100 cells in 3*MAX_INT microseconds from the receipt of this cell.
+3.1. Padding Machines
 
-This scheduled padding is non-periodic. For any forms of periodic
-padding, implementations should use the RELAY_COMMAND_PADDING_ADAPTIVE
-cell from Section 3.2 instead.
+Circuits can have either one or two state machines at both the origin and at a
+specified middle hop.
 
-3.2. Adaptive Padding message (RELAY_COMMAND_PADDING_ADAPTIVE)
+Each state machine can contain up to three states ("Start", "Burst" and
+"Gap") governing their behavior. Not all states need to be used.
 
-The following message is a generalization of the Adaptive Padding
-defense specified in "Timing Attacks and Defenses"[2].
+Each state of a padding machine specifies either:
+  * A histogram describing inter-arrival cell delays; OR
+  * A parameterized delay probability distribution for inter-arrival cell delays
 
-The message encodes either one or two state machines, each of which can
-contain one or two histograms ("Burst" and "Gap") governing their
-behavior.
-
-The "Burst" histogram specifies the delay probabilities for sending a
-padding packet after the arrival of a non-padding data packet.
+In either case, the lower bound of the delay probability distribution can be
+specified as a parameter, or it can be learned by measuring the RTT of the
+circuit.
```

Comment:
How does this RTT trick work exactly? And how do we choose which behavior to do?

Reply:
Should we specify how the RTT works in the proposal? It basically measures the time between "recieved" and "sent" non-padding cells at the middle node, until we see more than one of either back-to-back.

Reply:
Specified RTT calculation in https://github.com/torproject/torspec/pull/41/commits/1dd7f1ff78fcb69121130cbd802b0b8b527ffc63